librosa/core/pitch.py

```diff
@@ -333,3 +338,446 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+    for i, yin_frame in enumerate(yin_frames.T):
+        # 2. For each frame find the troughs.
+        is_trough = util.localmax(-yin_frame, axis=0)
+        is_trough = is_trough[1:-1]
+        yin_frame = yin_frame[1:-1]
+        trough_index, = np.nonzero(is_trough)
+
+        if len(trough_index) == 0:
+            continue
+
+        # 3. Find the troughs below each threshold.
+        trough_heights = yin_frame[trough_index]
+        trough_thresholds = trough_heights[:, None] < thresholds[None, 1:]
+
+        # 4. Define the prior over the troughs.
+        # Smaller periods are weighted more.
+        trough_positions = np.cumsum(trough_thresholds, axis=0) - 1
+        n_troughs = np.count_nonzero(trough_thresholds, axis=0)
+        trough_prior = scipy.stats.boltzmann.pmf(
+            trough_positions, boltzmann_parameter, n_troughs)
+        trough_prior[~trough_thresholds] = 0
+
+        # 5. For each threshold add probability to global minimum if no trough is below threshold,
+        # else add probability to each trough below threshold biased by prior.
+        probs = np.sum(trough_prior * beta_probs, axis=1)
+        global_min = np.argmin(trough_heights)
+        n_thresholds_below_min = np.count_nonzero(~trough_thresholds[global_min, :])
+        probs[global_min] += no_trough_prob*np.sum(beta_probs[:n_thresholds_below_min])
+
+        yin_probs[trough_index, i] = probs
+
+    yin_period, frame_index = np.nonzero(yin_probs)
+
+    # Refine peak by parabolic interpolation.
+    period_candidates = min_period + yin_period
+    period_candidates = period_candidates + parabolic_shifts[yin_period, frame_index]
+    f0_candidates = sr / period_candidates
+
+    n_bins_per_semitone = math.ceil(1. / resolution)
+    n_pitch_bins = math.floor(12*n_bins_per_semitone*np.log2(fmax / fmin)) + 1
+
+    # Construct transition matrix.
+    max_semitones_per_frame = round(max_transition_rate * 12 * hop_length / sr)
+    transition_width = max_semitones_per_frame*n_bins_per_semitone + 1
+    # Construct the within voicing transition probabilities
+    transition = sequence.transition_local(
+        n_pitch_bins, transition_width, window='triangle', wrap=False)
+    # Include across voicing transition probabilities
+    transition = np.block(
+        [[(1-switch_prob)*transition, switch_prob*transition],
+         [switch_prob*transition, (1-switch_prob)*transition]])
+
+    # Find pitch bin corresponding to each f0 candidate.
+    bin_index = 12*n_bins_per_semitone*np.log2(f0_candidates / fmin)
+    bin_index = np.clip(np.round(bin_index), 0, n_pitch_bins).astype(int)
+
+    # Observation probabilities.
+    observation_probs = np.zeros((2*n_pitch_bins, yin_frames.shape[1]))
+    observation_probs[bin_index, frame_index] = yin_probs[yin_period, frame_index]
+    voiced_prob = np.clip(np.sum(observation_probs, axis=0), 0, 1)
+    observation_probs[n_pitch_bins:, :] = (1 - voiced_prob[None, :]) / n_pitch_bins
```

Comment:
I had an idea here that improves performance a bit (but differs from the paper) and wanted to know what you thought @bmcfee.
Here the total unvoiced probability is divided uniformly over the unvoiced bins. So if a sequence of frames is decoded as unvoiced the best guess f0 just interpolates between the adjacent voiced frames even if there is a non-negligible voiced probability. If instead, we used:
```suggestion
observation_probs[n_pitch_bins:, :] = 0.5*observation_probs[n_pitch_bins:, :] 
observation_probs[n_pitch_bins:, :] += (1 -0.5* voiced_prob[None, :]) / n_pitch_bins
```
Then the best guess f0 in unvoiced frames would be more likely to track f0 as if the frame had been voiced.

Reply:
I think this relates to my [previous question](https://github.com/librosa/librosa/pull/1063#issuecomment-641267444) about how we define the probabilities here.

The current implementation, if i'm reading this correctly, is deriving voiced frequency probability `P[f, v]` from `yin_probs[f]` (for each `f`), and puts a uniform distribution over the unvoiced states `P[f, ~v]`, and then uses the state transition matrix to encourage continuity in unvoiced regions.  (This happens by virtue of having a diagonal-block matrix for T and using viterbi; in unvoiced regions, there would be no incentive to deviate from the previous state if the obs probabilities there are uniform.)

What you're suggesting also makes sense, leaking some of the yin probabilities into the unvoiced state observations puts a little more incentive to follow possible f0 trajectories, even if the likelihood of voicing isn't enough to kick it into the detection range.

You say it "improves performance", but along which metric?  I expect this to improve tracking at the expense of voicing correctness.

I'm not sure about the exact way that you're proposing to implement this idea though; shouldn't the 0.5 in 781 go out front?

More generally, we could add a balancing parameter `alpha` here, so instead of

```python
observation_probs[n_pitch_bins:, :] = (1 - voiced_prob[None, :]) / n_pitch_bins
```
we'd have
```python
observation_probs[n_pitch_bins:, :] = alpha*observation_probs[n_pitch_bins:, :]
observation_probs[n_pitch_bins:, :] = (1 - alpha) * (1 - voiced_prob[None, :]) / n_pitch_bins
```
`alpha=0` preserves the original `pyin`, and `alpha>0` smoothly mixes the uniform and yin-based state probabilities.  This way we don't lose backward compatibility, but can still get a little improvement if it helps.

Reply:
> You say it "improves performance", but along which metric? I expect this to improve tracking at the expense of voicing correctness.

Yeah, I haven't evaluated the trade off at all. Just noticed saw a small improvement on raw pitch accuracy on a few tracks.

> I'm not sure about the exact way that you're proposing to implement this idea though; shouldn't the 0.5 in 781 go out front?

Working it out properly I think it should be:
```python
observation_probs[n_pitch_bins:, :] = alpha*observation_probs[n_pitch_bins:, :] + (1 - (1+alpha)*voiced_prob[None, :]) / n_pitch_bins
```
Since we want the sum for each frame to equal 1 i.e. the above should sum to `1 - voiced_prob`.

Reply:
Oh I see, got it. :+1:

Reply:
I actually got this wrong again :weary:. This: `(1 - (1+alpha)*voiced_prob[None, :])` can result in negative values if `voiced_prob` is high. Taking another shot... basically we want a convex combination of the uniform distribution and whatever the voiced distribution is. So:
```python
voiced_distribution = observation_probs[:n_pitch_bins, :]
voiced_distribution[:, voiced_prob > 0] /= voiced_prob[None, voiced_prob > 0]  # normalize
observation_probs[n_pitch_bins:, voiced_prob > 0] = alpha*voiced_distribution[:, voiced_prob>0] + (1 - alpha)/n_pitch_bins
observation_probs[n_pitch_bins:, voiced_prob == 0] = 1 / n_pitch_bins
observation_probs[n_pitch_bins:, :] *= (1 - voiced_prob[None, :])  # multipy so unvoiced probs sum to 1 - voiced_prob
```

I did a few tests and it seems to give about a 0.1% boost to raw pitch accuracy. Do you think it's worth the added complexity?

Reply:
> Taking another shot... basically we want a convex combination of the uniform distribution and whatever the voiced distribution is.

I think that's what I had initially suggested?



> Since we want the sum for each frame to equal 1

I'm not sure we want or need that here.  The pyin algorithm, for better or worse, uses the viterbi (generative) algorithm, not `viterbi_discriminative`.  In the former case, we have observations of the form `P[X | Y]` where `X` corresponds to data (eg autocorrelation frames) and `Y` corresponds to states (f0 x voicing), and viterbi is finding the argmax P[Yi] under markov assumptions.  Note that in this case, `P[X | Y]` need not sum to 1 over all choices of Y: multiple states can give high probability to an X, or maybe no states give high probability to an X; that's fine.

The discriminative case assumes we have observations `P[Y | X]` (probably a much more realistic model of what the algorithm should be doing, but that's beside the point).  In this case, we *would* need the sum over Y of `P[Y|X]` to be 1.  The underlying decoding algorithm works in pretty much the same way -- we actually just do some bayes fiddling to turn it into the generative case, and with a uniform marginal distribution for `P[Y]` it'll give the same answers -- but the semantics of the variables involved are different.

Anyway, this is all to say that I don't see any problem with keeping the algorithm as is (generative assumption) and doing the convex combination that I suggested initially.

Reply:
> Since we want the sum for each frame to equal 1
>> I'm not sure we want or need that here.

Yeah, you're 100% right! Sorry, I confused myself. I'll go ahead and implement your suggestion. Any thoughts on what to call the `alpha` parameter?

Reply:
Maybe something like `unvoiced_weight`?

I also think I had a typo in my previous post, it should be
```python
observation_probs[n_pitch_bins:, :] = alpha*observation_probs[:n_pitch_bins, :]
observation_probs[n_pitch_bins:, :] = (1 - alpha) * (1 - voiced_prob[None, :]) / n_pitch_bins
```
(first line had the `n_pitch_bins:` on the RHS, which would be a no-op)

Reply:
So, without normalizing the first frame always gets decoded as unvoiced in the tests. Since the initial probability is only over unvoiced states and `alpha*observation_probs[:n_pitch_bins, :]` adds a largish value to an unvoiced state. Do you think this is something to be concerned about or should I just set `alpha=0` or `fill_na=None` in the tests?

Reply:
Hrm. Maybe we should just back off of this change for now and stick to the original method?

We can always add unvoiced tracking back in later, but I think it will take a bit more careful thinking and I don't want to hold this up unnecessarily.

Reply:
Yeah, agree with that. 

Not sure if any other changes are required? I think everything has been addressed so far?

Reply:
Great!  I'll give it one more read-through later on today, and hopefully we can get this merged.